aws-cpp-sdk-core/source/Aws.cpp

```diff
@@ -111,4 +111,45 @@ namespace Aws
         }
 #endif // USE_AWS_MEMORY_MANAGEMENT
     }
+
+    std::atomic<unsigned> APIWrapper::RefCount = 0;
+    std::mutex APIWrapper::InitGuard;
+
+    APIWrapper::APIWrapper(const SDKOptions& options)
+    {
+        if (RefCount.load() == 0u)
+        {
+            std::lock_guard<std::mutex> locker(InitGuard);
+
+            if (RefCount.load() == 0u)
```

Comment:
Hi. I think the separated if test outside the mutex makes the logic not thread-safe. I think a fail case is the following. This is on some computer with 8 CPU cores and a multithreading OS capable of simultaneous thread execution + time-slicing.

1. Two threads have previously run the constructor. Therefore, the value of RefCount = 2.
2. Thread 1 runs the destructor. On line 140, it evals to false because the value of RefCount is 2. Therefore, this thread will run the else clause *sometime* in the future.
3. Thread2 runs the destructor. On line 140, it evals to false because the value of RefCount is 2. Therefore, this thread will run the else clause sometime in the future.
4. Thread 1 runs line 152 and decrements RefCount. Its value is now 1.
5. Thread 2 runs line 152 and decrements RefCount. Its value is now 0.

This results in no thread calling Aws::ShutdownAPI(). You can also run the above in parallel by steps 2 and 3 running at exactly the same time. And steps 4 and 5 running at the same time. The value in RefCount will not be corrupted because it is atomic. However, the if-test was not atomic/protected and therefore fails these scenarios.

* You can remove the duplicate if tests in the contructor and destructor by using atomic.fetch_add/sub() to do an atomic read/math/write.
* I recommend not using <unsigned> and instead use <size_t>.  And you can make your SDKOptions static given the approach for this wrapper.

Perhaps this code is simpler yet safe. Comments?

```cpp
std::atomic<size_t> APIWrapper::s_count{ 0 };
std::mutex APIWrapper::s_initGuard{};
Aws::SDKOptions APIWrapper::s_awsOptions{};

APIWrapper::APIWrapper() {
    const size_t origCount = s_count.fetch_add(1);
    std::lock_guard<std::mutex> lock(s_initGuard);
    if (0 == origCount) {
        Aws::InitAPI(s_awsOptions);
    }    
}

APIWrapper::~APIWrapper() {
    const size_t origCount = s_count.fetch_sub(1);
    std::lock_guard<std::mutex> lock(s_initGuard);
    if (1 == origCount) {
        Aws::ShutdownAPI(s_awsOptions);
    }
}
```